DOCS-15702 Static Unlike Topologies (#101)
mongodb/docs-cluster-to-cluster-sync: source/reference/api/start.txt

       
       Default value is ``false``.

   * - ``sharding``
     - document
     - Optional
     - Configure synchronization between a replica set and sharded cluster.
       Synchronization from a replica set to a sharded cluster requires this
       option.

>        See :ref:`c2c-api-start-sharding` for more information.

Reviewer: [suggestion, non-blocking]

Our [style guide](https://www.mongodb.com/docs/meta/style-guide/style/links-and-cross-references/#make-link-text-clear) states that links should begin with the purpose of the link, rather than the link itself.

Suggest swapping the order here to:

For more information, see :ref:`c2c-api-start-sharding`.

Author: Fixed

       Default value is ``false``.

.. _c2c-api-start-sharding:

> Sharding Parameters

Reviewer: [question]

Would "fields" be a more appropriate name than "parameters" for these options? My understanding is that these are subfields of the `sharding` object.

Author: Fixed

Reviewer: Acknowledged. In that case, I'm fine leaving this as-is. Thanks for clarifying.

~~~~~~~~~~~~~~~~~~~

Synchronization from a replica set to a sharded cluster requires the 

> ``sharding`` option to shard collections on the destination cluster.

Reviewer: [suggestion]

Active voice might improve clarity here. Perhaps something like:

To synchronize from a replica set to a sharded cluster, you must set the `sharding` option.

Author: Fixed

> The ``sharding`` option has the following parameters:

Reviewer: Same question as above, should these be referred to as "fields"?

.. list-table::
   :header-rows: 1
   :stub-columns: 1

>    :widths: 15 15 70

Reviewer: [question]

Are all of the sharding subfields required? (assuming I'm specifying a sharding object).

Author: Fixed.

   * - Parameter
     - Type
     - Description

   * - ``createSupportedIndexes``
     - boolean

>      - Sets whether synchronization creates a supporting index 

Reviewer: [question]

Should we standardize on either "sets" or "specifies" in these descriptions?

Author: Fixed

       for the shard key, if none exists.

   * - ``shardingEntries``

>      - array

Reviewer: [question]

Would it be helpful to specify that this is an array of objects?

Author: Fixed

       Default value is ``false``.

.. _c2c-api-start-sharding:

Sharding Parameters
~~~~~~~~~~~~~~~~~~~

To synchronize from a replica set to a sharded cluster, set the 
``sharding`` option to shard collections on the destination cluster.

The ``sharding`` option has the following parameters:

.. list-table::
   :header-rows: 1
   :stub-columns: 1
   :widths: 15 15 70

   * - Parameter
     - Type
     - Description

   * - ``createSupportedIndexes``
     - boolean
     - Optional. Sets whether synchronization creates a supporting index 
       for the shard key, if none exists.  Defaults to ``false``.

   * - ``shardingEntries``
     - array of documents

>      - Required. Sets the namespace and key of collections to shard 

Reviewer 2: The array is required but can be empty, is it something that needs to be mentioned or not ?

Author: Fixed

       during synchronization.

>        Any collections not included in this array synchronize to unsharded 

Reviewer 2: I am not an English native speaker but isn't "any collection" (singular) more common ?

Author: Fixed

       collections on the destination cluster.

   * - | ``shardingEntries``
       | ``.collection``
     - string
     - Sets the collection to shard. 

   * - | ``shardingEntries``
       | ``.database``
     - string
     - Sets the database of the collection to shard.
 
   * - | ``shardingEntries``
       | ``.shardCollection``
     - document 
     - Sets the shard key to generate on the destination cluster.

   * - | ``shardingEntries``
       | ``.shardCollection``
       | ``.key``
     - array 

>      - Sets the indexes to use for the shard key. 

Reviewer 2: I think it should say "Sets the **fields** to use for the shard key.". Also it might be too long for here but should it say somewhere that the rules are the same as for shardCollection (in term of supporting up to one `"hashed"` field and 0-N `"1"` range fields) ?

Author: Fixed

Response
~~~~~~~~

.. literalinclude:: /includes/api/responses/success.json
   :language: json
   :copyable: false

Example 3 - Start Sync from Replica Set to Sharded Cluster
----------------------------------------------------------

Request
~~~~~~~

.. literalinclude:: /includes/api/requests/start-rs-shard.sh
   :language: shell

Response
~~~~~~~~

.. literalinclude:: /includes/api/responses/success.json
   :language: json
   :copyable: false

Behavior
--------

State
~~~~~

If the ``start`` request is successful, ``mongosync`` enters the
``RUNNING`` state.

Shard Collections
~~~~~~~~~~~~~~~~~~~~~

> Synchronization from a replica set to a sharded cluster requires the 

Author: @jeff-allen-mongo: Good catch.  SC > RC is not supported (it throws an error), but I clipped it during edits.

The `sharding` option is required for RC > SC, since it needs special configuration to know which collections to shard and which keys to use in sharding.  Asymmetric SC > SC does not require special configuration, it just works on its own.

> Synchronization from a replica set to a sharded cluster requires the 

Reviewer: Got it. Thanks for the context.

``sharding`` option, to specify how ``mongosync`` should shard collections.

The ``sharding.shardEntries`` array specifies the collections to shard.
Collections that are not listed in this array replicate as unsharded.

Supporting Indexes
~~~~~~~~~~~~~~~~~~

Synchronization from a replica set to a sharded cluster can automatically
create a supporting index for the new shard key on the destination cluster.
This is done by setting the ``sharding.createSupportingIndexes`` option.

> When ``sharding.createSupportingIndexes`` is ``false`` (the default): 

Reviewer: Should we mention in the table above that the default value is `false`?

Author: Fixed

> * Each shard key the user provides for the ``sharding.shardEntries`` option

Author: Fixed

  must have an existing index on the source cluster.

* One of the indexes used for the shard key must have simple collation if the 
  collection uses any other collation.

* To use a unique index in the shard key, you must specify its uniqueness when
  you create the index on the source cluster. 

* Unique indexes on the source cluster that are incompatible with the requested
  shard key on the destination cluster, such as a unique index on the source
  that does not contain the requested shard key as a prefix on the destination,
  can cause ``mongosync`` to fail.

When ``sharding.createSupportingIndexes`` is ``true``: 

* ``mongosync`` synchronizes indexes on the source cluster to the destination.

* If the supporting indexes exist on the source cluster, ``mongosync`` 
  synchronizes the indexes to the destination cluster and uses them 
  as the shard key.

* If the supporting indexes don't exist, ``mongosync`` creates them on the
  destination cluster.

> ``mongosync`` either creates supporting indexes for all collections or none.

Author: Fixed

> Rename during Sync

Reviewer: I think "During" should be capitalized here.

```suggestion
Rename During Sync
```

Author: Fixed

~~~~~~~~~~~~~~~~~~

Collections synchronized from a replica set to a sharded cluster listed in

> the ``sharding.shardEntries`` array, become sharded collections on the

Reviewer: not sure we need the comma after array?

Author: Fixed

destination cluster.

> Renaming a collection, such as with the :dbcommand:`renameCollection` command

Reviewer: I'm having a little trouble parsing this bullet. Could we write this as:

Renaming a collection (such as with the :dbcommand:`renameCollection` command) on the source cluster...

Author: Fixed

on the source cluster after calling ``start`` but before ``mongosync`` begins 
to copy the collection can block the collection from sharding on the destination.

To check whether a collection is safe to rename, call the :ref:`c2c-api-progress`
endpoint and check the value of the ``collectionCopy.estimatedCopiedBytes`` field 
in the return document.  

> * A value of 0 indicates that ``mongosync`` has not yet begun to copy the

Reviewer: [minimalism]

```suggestion
* A value of 0 indicates that ``mongosync`` has not started to copy the
```

Author: Fixed

  collection.  Renaming the collection at this point results in the collection 
  being unsharded on the destination cluster.

> * A value greater than 0 indicates that ``mongosync`` has begun the copy and

Reviewer: [for consistency with the previous suggestion]

```suggestion
* A value greater than 0 indicates that ``mongosync`` has started the copy and
```

Author: Fixed

>   read the shard map.  Renaming the collection from this point on does not block

Reviewer: I'm not sure what "read the shard map" means.

Author: Fixed

Response
~~~~~~~~

.. literalinclude:: /includes/api/responses/success.json
   :language: json
   :copyable: false

Example 3 - Start Sync from Replica Set to Sharded Cluster
----------------------------------------------------------

Request
~~~~~~~

.. literalinclude:: /includes/api/requests/start-rs-shard.sh
   :language: shell

Response
~~~~~~~~

.. literalinclude:: /includes/api/responses/success.json
   :language: json
   :copyable: false

Behavior
--------

State
~~~~~

If the ``start`` request is successful, ``mongosync`` enters the
``RUNNING`` state.

Shard Collections
~~~~~~~~~~~~~~~~~~~~~

Synchronization from a replica set to a sharded cluster requires the 

> ``sharding`` option, to specify how ``mongosync`` should shard collections.

Reviewer: [suggestion]

IMO, breaking this into two separate sentences and using a similar second-person voice to the edit we made above would make this easier to understand. Consider something like:

"To synchronize a replica set to a sharded cluster, specify the ``sharding`` option. The ``sharding option`` specifies how ``mongosync`` shards collections."

Author: Fixed

* One of the indexes used for the shard key must have simple collation if the 
  collection uses any other collation.

* To use a unique index in the shard key, you must specify its uniqueness when
  you create the index on the source cluster. 

* Unique indexes on the source cluster that are incompatible with the requested
  shard key on the destination cluster, such as a unique index on the source
  that does not contain the requested shard key as a prefix on the destination,
  can cause ``mongosync`` to fail.

When ``sharding.createSupportingIndexes`` is ``true``: 

* ``mongosync`` synchronizes indexes on the source cluster to the destination.

* If the supporting indexes exist on the source cluster, ``mongosync`` 
  synchronizes the indexes to the destination cluster and uses them 
  as the shard key.

* If the supporting indexes don't exist, ``mongosync`` creates them on the
  destination cluster.

The ``sharding.createSupportingIndexes`` option affects all collections sharded
during sync.

Rename During Sync
~~~~~~~~~~~~~~~~~~

Collections synchronized from a replica set to a sharded cluster listed in
the ``sharding.shardEntries`` array become sharded collections on the

> destination cluster.

Reviewer: [question, minimalism]

Do we need to say "listed in the ``sharding.shardEntries`` array" in this sentence? The sentence makes sense to me without that phrase, so removing it might make this a bit clearer.

Author: I reworked it a bit, but I think we need the "listed" clause to drive home that collections not in that list remain unsharded.

collections.

The ``sharding.shardEntries`` array specifies the collections to shard.
Collections that are not listed in this array replicate as unsharded.

Supporting Indexes
~~~~~~~~~~~~~~~~~~

Synchronization from a replica set to a sharded cluster can automatically
create a supporting index for the new shard key on the destination cluster.
This is done by setting the ``sharding.createSupportingIndexes`` option.

When ``sharding.createSupportingIndexes`` is ``false`` (the default): 

* Each shard key you provide for the ``sharding.shardEntries`` option
  must have an existing index on the source cluster.

* One of the indexes used for the shard key must have simple collation if the 
  collection uses any other collation.

* To use a unique index in the shard key, you must specify its uniqueness when
  you create the index on the source cluster. 

* Unique indexes on the source cluster that are incompatible with the requested
  shard key on the destination cluster, such as a unique index on the source
  that does not contain the requested shard key as a prefix on the destination,
  can cause ``mongosync`` to fail.

When ``sharding.createSupportingIndexes`` is ``true``: 

> * ``mongosync`` synchronizes indexes on the source cluster to the destination.

Reviewer 2: I find it a bit confusing for it to be here, this is true in both cases (`sharding.createSupportingIndexes = false` too). Maybe it should be before the 2 "When ``sharding.createSupportingIndexes``" paragraphs ?

Author: Fixed

* If the supporting indexes exist on the source cluster, ``mongosync`` 
  synchronizes the indexes to the destination cluster and uses them 
  as the shard key.

* If the supporting indexes don't exist, ``mongosync`` creates them on the
  destination cluster.

The ``sharding.createSupportingIndexes`` option affects all collections sharded
during sync.

Rename During Sync
~~~~~~~~~~~~~~~~~~

Collections listed in the ``sharding.shardEntries`` array 
when synchronized from the replica set to a sharded cluster 
become sharded collections on the destination cluster.

Renaming a collection (such as with the :dbcommand:`renameCollection` command)
on the source cluster after calling ``start`` but before ``mongosync`` begins 
to copy the collection can block the collection from sharding on the destination.

> To check whether a collection is safe to rename, call the :ref:`c2c-api-progress`

Reviewer 2: Technically this is not per collection: progress will return > 0 only once it is safe for **all collections**.

Author: Fixed

endpoint and check the value of the ``collectionCopy.estimatedCopiedBytes`` field 
in the return document.  

> * A value of 0 indicates that ``mongosync`` has not started to copy the

Reviewer 2: The > 0 is correct but technically if they see 0, it is racy: by the time the user calls rename on source, maybe mongosync progressed a bit and calling `progress` again would return > 0. I don't know how to write this clearly but the idea is:

A value of 0 indicates that ``mongosync`` has not started to copy collections. Renaming a collection at this point might result in the collection being unsharded on the destination cluster: Mongosync transition to copying collections can happen before or after the rename takes effect on source.

Author: Fixed

   :language: shell

Response
~~~~~~~~

.. literalinclude:: /includes/api/responses/success.json
   :language: json
   :copyable: false

Behavior
--------

State
~~~~~

If the ``start`` request is successful, ``mongosync`` enters the
``RUNNING`` state.

Pre-Split Chunks
~~~~~~~~~~~~~~~~

.. versionadded:: 1.1

When ``mongosync`` syncs to a sharded cluster, it pre-splits chunks for
sharded collections on the destination cluster.  This is supported in the
following configurations:

* Sync from a replica set to a sharded cluster.

* Sync between sharded clusters that differ in the number of shards.

Shard Replica Sets 
~~~~~~~~~~~~~~~~~~

.. versionadded:: 1.1

Sync from a replica set to a sharded cluster requires the 
``sharding`` option. This option configures how ``mongosync`` shards
collections.

The ``sharding.shardEntries`` array specifies the collections to shard.
Collections that are not listed in this array replicate as unsharded.

Supporting Indexes
~~~~~~~~~~~~~~~~~~

.. versionadded:: 1.1

``mongosync`` syncs indexes from the source cluster to the destination
cluster.  But, when syncing from a replica set to a sharded cluster,

> ``mongosync`` may require the an additional index to support the shard key,

Reviewer 2: Extra "the" before "an".

Author: Fixed

which may not exist on the source cluster.

``mongosync`` can create supporting indexes for sharded collections during sync.
This is done by setting the ``sharding.createSupportingIndexes`` option.

When ``sharding.createSupportingIndexes`` is ``false`` (the default): 

* Each shard key you provide for the ``sharding.shardEntries`` option
  must have an existing index on the source cluster.

* One of the indexes used for the shard key must have simple collation if the 
  collection uses any other collation.

* To use a unique index in the shard key, you must specify its uniqueness when
  you create the index on the source cluster. 

* Unique indexes on the source cluster that are incompatible with the requested
  shard key on the destination cluster, such as a unique index on the source
  that does not contain the requested shard key as a prefix on the destination,
  can cause ``mongosync`` to fail.

When ``sharding.createSupportingIndexes`` is ``true``: 

* If the supporting indexes exist on the source cluster, ``mongosync`` 
  syncs the indexes to the destination cluster and uses them 

>   as the shard key.

Reviewer 2: The first part of the sentence is plural (indexes) but the second is singular. Maybe just "If the supporting indexes exist on the source cluster, ``mongosync`` syncs the indexes to the destination cluster and uses them" ?

Author: Refactored.

* If the supporting indexes don't exist, ``mongosync`` creates them on the
  destination cluster.

> The ``sharding.createSupportingIndexes`` option affects all collections sharded

Reviewer 2: I would omit "during sync" (as it sounds that the collections are being sharded during Mongosync run) and say "affects sharded all collections".

Author: Fixed

during sync.

Rename During Sync
~~~~~~~~~~~~~~~~~~

.. versionadded:: 1.1

Collections listed in the ``sharding.shardEntries`` array 

> when synced from the replica set to a sharded cluster 

Reviewer 2: I think "from a replica set" is better here.

Author: Fixed

become sharded collections on the destination cluster.

Renaming a collection (such as with the :dbcommand:`renameCollection` command)
on the source cluster after calling ``start`` but before ``mongosync`` begins 
to copy the collection can block the collection from sharding on the destination.

.. note:: 

   Renaming collections to use a different database while ``mongosync``

>    is running is not supported

Reviewer 2: I would add "when syncing from a replica set to a sharded cluster" in the note too.

Author: Fixed.

* If the supporting indexes exist on the source cluster, ``mongosync`` 
  syncs the indexes to the destination cluster and uses them 
  as shard keys. 

* If the supporting indexes don't exist, ``mongosync`` creates them on the
  destination cluster.

> The ``sharding.createSupportingIndexes`` option affects all collections sharded.

Reviewer 2: Nit, I would change `option affects all collections sharded.` to `option affects all sharded collections.`

Author: Fixed
